File: cpp/lexer.cpp
#include "lexer.hpp"
#include <cctype>
#include <unordered_map>

namespace mpp {

static const std::unordered_map<std::string, TokenType> KEYWORDS = {
    {"let", TokenType::Let}, {"const", TokenType::Const}, {"fn", TokenType::Fn},
    {"if", TokenType::If}, {"elif", TokenType::Elif}, {"else", TokenType::Else},
    {"for", TokenType::For}, {"while", TokenType::While}, {"in", TokenType::In},
    {"return", TokenType::Return}, {"class", TokenType::Class},
    {"pub", TokenType::Pub}, {"priv", TokenType::Priv}, {"self", TokenType::Self},
    {"new", TokenType::New}, {"true", TokenType::True}, {"false", TokenType::False},
};

Lexer::Lexer(const std::string& source) : source_(source) {}

char Lexer::current() const {
    return pos_ >= source_.size() ? '\0' : source_[pos_];
}

char Lexer::peek(size_t off) const {
    size_t p = pos_ + off;
    return p >= source_.size() ? '\0' : source_[p];
}

char Lexer::advance() {
    if (pos_ >= source_.size()) return '\0';
    char ch = source_[pos_++];
    if (ch == '\n') { line_++; column_ = 1; }
    else column_++;
    return ch;
}

void Lexer::skipWhitespace() {
    while (current() == ' ' || current() == '\t' || current() == '\n' || current() == '\r')
        advance();
}

void Lexer::skipComment() {
    if (current() == '/' && peek() == '/')
        while (current() != '\n' && current() != '\0') advance();
}

Token Lexer::readNumber() {
    int sl = line_, sc = column_;
    std::string s;
    while (std::isdigit(current()) || current() == '.')
        s += advance();
    Token t;
    t.type = TokenType::Number;
    t.value = s;
    t.line = sl; t.column = sc;
    return t;
}

Token Lexer::readString(char quote) {
    int sl = line_, sc = column_;
    advance();
    std::string s;
    while (current() != quote && current() != '\0') {
        if (current() == '\\') {
            advance();
            char e = advance();
            if (e == 'n') s += '\n';
            else if (e == 't') s += '\t';
            else if (e == '"' || e == '\'') s += e;
            else s += e;
        } else s += advance();
    }
    if (current() == quote) advance();
    Token t;
    t.type = TokenType::String;
    t.value = s;
    t.line = sl; t.column = sc;
    return t;
}

Token Lexer::readIdent() {
    int sl = line_, sc = column_;
    std::string s;
    while (std::isalnum(static_cast<unsigned char>(current())) || current() == '_')
        s += advance();
    Token t;
    auto it = KEYWORDS.find(s);
    t.type = it != KEYWORDS.end() ? it->second : TokenType::Ident;
    t.value = s;
    t.line = sl; t.column = sc;
    return t;
}

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> tokens;
    while (pos_ < source_.size()) {
        skipWhitespace();
        if (pos_ >= source_.size()) break;
        skipComment();
        if (pos_ >= source_.size()) break;
        char ch = current();
        int sl = line_, sc = column_;

        if (ch == '=' && peek() == '=') { advance(); advance(); tokens.push_back({TokenType::EqEq, "==", sl, sc}); continue; }
        if (ch == '!' && peek() == '=') { advance(); advance(); tokens.push_back({TokenType::Neq, "!=", sl, sc}); continue; }
        if (ch == '<' && peek() == '=') { advance(); advance(); tokens.push_back({TokenType::Le, "<=", sl, sc}); continue; }
        if (ch == '>' && peek() == '=') { advance(); advance(); tokens.push_back({TokenType::Ge, ">=", sl, sc}); continue; }
        if (ch == '-' && peek() == '>') { advance(); advance(); tokens.push_back({TokenType::Arrow, "=>", sl, sc}); continue; }
        if (ch == '=' && peek() == '>') { advance(); advance(); tokens.push_back({TokenType::Arrow, "=>", sl, sc}); continue; }
        if (ch == '&' && peek() == '&') { advance(); advance(); tokens.push_back({TokenType::And, "&&", sl, sc}); continue; }
        if (ch == '|' && peek() == '|') { advance(); advance(); tokens.push_back({TokenType::Or, "||", sl, sc}); continue; }

        if (ch == '+') { advance(); tokens.push_back({TokenType::Plus, "+", sl, sc}); }
        else if (ch == '-') { advance(); tokens.push_back({TokenType::Minus, "-", sl, sc}); }
        else if (ch == '*') { advance(); tokens.push_back({TokenType::Star, "*", sl, sc}); }
        else if (ch == '/') { advance(); tokens.push_back({TokenType::Slash, "/", sl, sc}); }
        else if (ch == '%') { advance(); tokens.push_back({TokenType::Percent, "%", sl, sc}); }
        else if (ch == '=') { advance(); tokens.push_back({TokenType::Eq, "=", sl, sc}); }
        else if (ch == '<') { advance(); tokens.push_back({TokenType::Lt, "<", sl, sc}); }
        else if (ch == '>') { advance(); tokens.push_back({TokenType::Gt, ">", sl, sc}); }
        else if (ch == '!') { advance(); tokens.push_back({TokenType::Not, "!", sl, sc}); }
        else if (ch == '(') { advance(); tokens.push_back({TokenType::LParen, "(", sl, sc}); }
        else if (ch == ')') { advance(); tokens.push_back({TokenType::RParen, ")", sl, sc}); }
        else if (ch == '{') { advance(); tokens.push_back({TokenType::LBrace, "{", sl, sc}); }
        else if (ch == '}') { advance(); tokens.push_back({TokenType::RBrace, "}", sl, sc}); }
        else if (ch == '[') { advance(); tokens.push_back({TokenType::LBracket, "[", sl, sc}); }
        else if (ch == ']') { advance(); tokens.push_back({TokenType::RBracket, "]", sl, sc}); }
        else if (ch == ',') { advance(); tokens.push_back({TokenType::Comma, ",", sl, sc}); }
        else if (ch == '.') { advance(); tokens.push_back({TokenType::Dot, ".", sl, sc}); }
        else if (ch == ':') { advance(); tokens.push_back({TokenType::Colon, ":", sl, sc}); }
        else if (ch == ';') { advance(); tokens.push_back({TokenType::Semicolon, ";", sl, sc}); }
        else if (ch == '"' || ch == '\'') tokens.push_back(readString(ch));
        else if (std::isdigit(static_cast<unsigned char>(ch))) tokens.push_back(readNumber());
        else if (std::isalpha(static_cast<unsigned char>(ch)) || ch == '_') tokens.push_back(readIdent());
        else advance();
    }
    tokens.push_back({TokenType::Eof, "", line_, column_});
    return tokens;
}

} // namespace mpp

File: mpp/lexer.py
"""
M++ Lexer — Tokenizes M++ source code
"""
import re
from dataclasses import dataclass
from enum import Enum
from typing import List, Optional


class TokenType(Enum):
    # Literals
    NUMBER = "NUMBER"
    STRING = "STRING"
    IDENT = "IDENT"
    TRUE = "TRUE"
    FALSE = "FALSE"

    # Keywords
    LET = "LET"
    CONST = "CONST"
    FN = "FN"
    IF = "IF"
    ELIF = "ELIF"
    ELSE = "ELSE"
    FOR = "FOR"
    WHILE = "WHILE"
    IN = "IN"
    RETURN = "RETURN"
    CLASS = "CLASS"
    PUB = "PUB"
    PRIV = "PRIV"
    SELF = "SELF"
    NEW = "NEW"

    # Operators
    PLUS = "PLUS"
    MINUS = "MINUS"
    STAR = "STAR"
    SLASH = "SLASH"
    PERCENT = "PERCENT"
    EQ = "EQ"
    EQEQ = "EQEQ"
    NEQ = "NEQ"
    LT = "LT"
    LE = "LE"
    GT = "GT"
    GE = "GE"
    AND = "AND"
    OR = "OR"
    NOT = "NOT"
    ARROW = "ARROW"

    # Delimiters
    LPAREN = "LPAREN"
    RPAREN = "RPAREN"
    LBRACE = "LBRACE"
    RBRACE = "RBRACE"
    LBRACKET = "LBRACKET"
    RBRACKET = "RBRACKET"
    COMMA = "COMMA"
    DOT = "DOT"
    COLON = "COLON"
    SEMICOLON = "SEMICOLON"

    EOF = "EOF"


@dataclass
class Token:
    type: TokenType
    value: any
    line: int
    column: int


KEYWORDS = {
    "let": TokenType.LET,
    "const": TokenType.CONST,
    "fn": TokenType.FN,
    "if": TokenType.IF,
    "elif": TokenType.ELIF,
    "else": TokenType.ELSE,
    "for": TokenType.FOR,
    "while": TokenType.WHILE,
    "in": TokenType.IN,
    "return": TokenType.RETURN,
    "class": TokenType.CLASS,
    "pub": TokenType.PUB,
    "priv": TokenType.PRIV,
    "self": TokenType.SELF,
    "new": TokenType.NEW,
    "true": TokenType.TRUE,
    "false": TokenType.FALSE,
}


class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1
        self.column = 1

    def current(self) -> str:
        if self.pos >= len(self.source):
            return "\0"
        return self.source[self.pos]

    def peek(self, offset: int = 1) -> str:
        p = self.pos + offset
        if p >= len(self.source):
            return "\0"
        return self.source[p]

    def advance(self) -> str:
        if self.pos >= len(self.source):
            return "\0"
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.column = 1
        else:
            self.column += 1
        return ch

    def skip_whitespace(self):
        while self.current() in " \t\n\r":
            self.advance()

    def skip_comment(self):
        if self.current() == "/" and self.peek() == "/":
            while self.current() != "\n" and self.current() != "\0":
                self.advance()

    def read_number(self) -> Token:
        start_line, start_col = self.line, self.column
        num_str = ""
        while self.current().isdigit() or self.current() == ".":
            num_str += self.advance()
        val = float(num_str) if "." in num_str else int(num_str)
        return Token(TokenType.NUMBER, val, start_line, start_col)

    def read_string(self, quote: str) -> Token:
        start_line, start_col = self.line, self.column
        self.advance()  # consume opening quote
        s = ""
        while self.current() != quote and self.current() != "\0":
            if self.current() == "\\":
                self.advance()
                esc = self.advance()
                s += {"n": "\n", "t": "\t", '"': '"', "'": "'", "\\": "\\"}.get(esc, esc)
            else:
                s += self.advance()
        if self.current() == quote:
            self.advance()
        return Token(TokenType.STRING, s, start_line, start_col)

    def read_ident(self) -> Token:
        start_line, start_col = self.line, self.column
        ident = ""
        while self.current().isalnum() or self.current() == "_":
            ident += self.advance()
        tok_type = KEYWORDS.get(ident, TokenType.IDENT)
        return Token(tok_type, ident, start_line, start_col)

    def tokenize(self) -> List[Token]:
        tokens = []
        while self.pos < len(self.source):
            self.skip_whitespace()
            if self.pos >= len(self.source):
                break
            self.skip_comment()
            if self.pos >= len(self.source):
                break
            ch = self.current()
            start_line, start_col = self.line, self.column

            # Two-char operators
            if ch == "=" and self.peek() == "=":
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.EQEQ, "==", start_line, start_col))
                continue
            if ch == "!" and self.peek() == "=":
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.NEQ, "!=", start_line, start_col))
                continue
            if ch == "<" and self.peek() == "=":
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.LE, "<=", start_line, start_col))
                continue
            if ch == ">" and self.peek() == "=":
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.GE, ">=", start_line, start_col))
                continue
            if ch == "-" and self.peek() == ">":
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.ARROW, "=>", start_line, start_col))
                continue
            if ch == "=" and self.peek() == ">":
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.ARROW, "=>", start_line, start_col))
                continue
            if ch == "&" and self.peek() == "&":
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.AND, "&&", start_line, start_col))
                continue
            if ch == "|" and self.peek() == "|":
                self.advance()
                self.advance()
                tokens.append(Token(TokenType.OR, "||", start_line, start_col))
                continue

            # Single char
            if ch == "+":
                tokens.append(Token(TokenType.PLUS, ch, start_line, start_col))
                self.advance()
            elif ch == "-":
                tokens.append(Token(TokenType.MINUS, ch, start_line, start_col))
                self.advance()
            elif ch == "*":
                tokens.append(Token(TokenType.STAR, ch, start_line, start_col))
                self.advance()
            elif ch == "/":
                tokens.append(Token(TokenType.SLASH, ch, start_line, start_col))
                self.advance()
            elif ch == "%":
                tokens.append(Token(TokenType.PERCENT, ch, start_line, start_col))
                self.advance()
            elif ch == "=":
                tokens.append(Token(TokenType.EQ, ch, start_line, start_col))
                self.advance()
            elif ch == "<":
                tokens.append(Token(TokenType.LT, ch, start_line, start_col))
                self.advance()
            elif ch == ">":
                tokens.append(Token(TokenType.GT, ch, start_line, start_col))
                self.advance()
            elif ch == "!":
                tokens.append(Token(TokenType.NOT, ch, start_line, start_col))
                self.advance()
            elif ch == "(":
                tokens.append(Token(TokenType.LPAREN, ch, start_line, start_col))
                self.advance()
            elif ch == ")":
                tokens.append(Token(TokenType.RPAREN, ch, start_line, start_col))
                self.advance()
            elif ch == "{":
                tokens.append(Token(TokenType.LBRACE, ch, start_line, start_col))
                self.advance()
            elif ch == "}":
                tokens.append(Token(TokenType.RBRACE, ch, start_line, start_col))
                self.advance()
            elif ch == "[":
                tokens.append(Token(TokenType.LBRACKET, ch, start_line, start_col))
                self.advance()
            elif ch == "]":
                tokens.append(Token(TokenType.RBRACKET, ch, start_line, start_col))
                self.advance()
            elif ch == ",":
                tokens.append(Token(TokenType.COMMA, ch, start_line, start_col))
                self.advance()
            elif ch == ".":
                tokens.append(Token(TokenType.DOT, ch, start_line, start_col))
                self.advance()
            elif ch == ":":
                tokens.append(Token(TokenType.COLON, ch, start_line, start_col))
                self.advance()
            elif ch == ";":
                tokens.append(Token(TokenType.SEMICOLON, ch, start_line, start_col))
                self.advance()
            elif ch in '"\'':
                tokens.append(self.read_string(ch))
            elif ch.isdigit():
                tokens.append(self.read_number())
            elif ch.isalpha() or ch == "_":
                tokens.append(self.read_ident())
            else:
                self.advance()  # skip unknown

        tokens.append(Token(TokenType.EOF, None, self.line, self.column))
        return tokens
